WIP: WASM: Enable reflection for types

yowl opened the pull request:
Fixes #5842 
Thanks for all the help on this.
I'm leaving this WIP as it contains 2 changes from @morganbr in `ILToWebAssemblyImporter.cs`: the funclets for try/catch/finally which is needed here; and some alignment fixes which is not strictly necessary for the WASM=1 output, but it was earlier in the branch and was useful for debugging.  Ideally these changes would be merged separately but this work is getting old and I wanted to push it along as the merge to `master` was hard enough now.  And with respect to the merge, the work here might be useful for Morgan as due to the overlapping time of the `ThrowIfNull` changes and these the merge is not straightforward.   Also maybe the other parts of this could be reviewed?

MichalStrehovsky commented on src/ILCompiler.Compiler/src/Compiler/CompilerTypeSystemContext.BoxedTypes.cs:
```diff
@@ -112,13 +112,19 @@ public MethodDesc GetUnboxingThunk(MethodDesc targetMethod, ModuleDesc ownerModu
                 InstantiatedType boxedType = boxedTypeDefinition.MakeInstantiatedType(owningType.Instantiation);
                 MethodDesc thunk = GetMethodForInstantiatedType(thunkDefinition, boxedType);
                 //TODO: this might be triggered by a struct that implements an interface with a generic method
+                if (thunk.HasInstantiation)
```
This could use a small cleanup. I would structure the code this way:

```csharp
MethodDesc thunk;
if (owningType != owningTypeDefinition)
{
  InstantiatedType boxedType = boxedTypeDefinition.MakeInstantiatedType(owningType.Instantiation);
  thunk = GetMethodForInstantiatedType(thunkDefinition, boxedType);
}
else
{
  thunk = thunkDefinition;
}

if (thunk.HasInstantiation)
  thunk = thunk.MakeInstantiatedMethod(targetMethod.Instantiation);

return thunk;
```

Avoids the duplicate `if (HasInstantiation)` logic.

MichalStrehovsky commented on src/ILCompiler.Compiler/src/Compiler/CompilerTypeSystemContext.BoxedTypes.cs:
```diff
@@ -556,12 +562,35 @@ public override MethodIL EmitIL()
 
-                // Call an instance method on the target valuetype
-                codeStream.Emit(ILOpcode.call, emit.NewToken(_targetMethod.InstantiateAsOpen()));
+//                codeStream.Emit(ILOpcode.call, emit.NewToken(_targetMethod.InstantiateAsOpen()));
```
Delete the commented out code?

MichalStrehovsky commented on src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/ModuleMetadataNode.cs:
```diff
@@ -39,7 +39,7 @@ public override IEnumerable<DependencyListEntry> GetStaticDependencies(NodeFacto
             dependencies.Add(factory.TypeMetadata(_module.GetGlobalModuleType()), "Global module type");
 
-            CustomAttributeBasedDependencyAlgorithm.AddDependenciesDueToCustomAttributes(ref dependencies, factory, (EcmaAssembly)_module);
+            if(_module is EcmaAssembly) CustomAttributeBasedDependencyAlgorithm.AddDependenciesDueToCustomAttributes(ref dependencies, factory, (EcmaAssembly)_module);
```
I think this will not be needed once the blocking policy is fixed up.

MichalStrehovsky commented on src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/TypeMetadataNode.cs:
```diff
@@ -36,23 +36,25 @@ public override IEnumerable<DependencyListEntry> GetStaticDependencies(NodeFacto
             DependencyList dependencies = new DependencyList();
 
-            CustomAttributeBasedDependencyAlgorithm.AddDependenciesDueToCustomAttributes(ref dependencies, factory, ((EcmaType)_type));
+            if(_type is EcmaType) CustomAttributeBasedDependencyAlgorithm.AddDependenciesDueToCustomAttributes(ref dependencies, factory, ((EcmaType)_type));
```
I think the if check won't be needed once the blocking policy is fixed up.

MichalStrehovsky commented on src/ILCompiler.WebAssembly/src/Compiler/WebAssemblyMetadataBlockingPolicy.cs:
```diff
@@ -0,0 +1,22 @@
+    {
+        public override bool IsBlocked(MetadataType type)
+        {
+            return false;
```
Changing this to `return !(type is EcmaType);` should get rid of some of the other problematic code I pointed to.

MichalStrehovsky commented on src/ILCompiler.WebAssembly/src/Compiler/WebAssemblyMetadataBlockingPolicy.cs:
```diff
@@ -0,0 +1,22 @@
+﻿using Internal.TypeSystem;
+
+namespace ILCompiler.Compiler
```
Could you make this `namespace ILCompiler`? We don't use `ILCompiler.Compiler` as a namespace.

MichalStrehovsky commented on src/ILCompiler.WebAssembly/src/Compiler/WebAssemblyMetadataBlockingPolicy.cs:
```diff
@@ -0,0 +1,22 @@
+﻿using Internal.TypeSystem;
```
Could you add the copyright header?

MichalStrehovsky commented on src/ILCompiler.WebAssembly/src/Compiler/DependencyAnalysis/WebAssemblyMethodCodeNode.cs:
```diff
@@ -59,8 +59,17 @@ public override IEnumerable<DependencyListEntry> GetStaticDependencies(NodeFacto
 
-            foreach (Object node in _dependencies)
-                dependencies.Add(node, "Wasm code ");
+            if (_method.HasInstantiation && !factory.CanNecessaryTypeSymbol(_method.OwningType))
```
I'm a bit worried about this line.

Is it possible the need for this if block comes from the metadata policy needing tweaks? Do we still need it after addressing the other feedback?

yowl replied:
With the policy change, and this reverted it hits the assert at https://github.com/dotnet/corert/blob/74897d4ab3eaff1141f8ac07287ccefbfbb5a352/src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/ExactMethodInstantiationsNode.cs#L85 with stack
```
I'll do some more investigation on this one as this came up in the dev and there were some comments here https://github.com/dotnet/corert/issues/5842#issuecomment-430605137 and in your reply which I'll go over.
         Assertion Failed

            at ILCompiler.DependencyAnalysis.ExactMethodInstantiationsNode.GetData(NodeFactory factory, Boolean relocsOnly) in D:\GitHub\corert\src\ILCompiler.Compiler\src\Compiler\DependencyAnalysis\ExactMethodInstantiationsNode.cs:line 85
            at ILCompiler.DependencyAnalysis.WebAssemblyObjectWriter.EmitObject(String objectFilePath, IEnumerable`1 nodes, NodeFactory factory, WebAssemblyCodegenCompilation compilation, IObjectDumper dumper) in D:\GitHub\corert\src\ILCompiler.WebAssembly\src\CodeGen\WebAssemblyObjectWriter.cs:line 772
            at ILCompiler.WebAssemblyCodegenCompilation.CompileInternal(String outputFile, ObjectDumper dumper) in D:\GitHub\corert\src\ILCompiler.WebAssembly\src\Compiler\WebAssemblyCodegenCompilation.cs:line 54
            at ILCompiler.Compilation.ILCompiler.ICompilation.Compile(String outputFile, ObjectDumper dumper) in D:\GitHub\corert\src\ILCompiler.Compiler\src\Compiler\Compilation.cs:line 387
            at ILCompiler.Program.Run(String[] args) in D:\GitHub\corert\src\ILCompiler\src\Program.cs:line 576
            at ILCompiler.Program.Main(String[] args) in D:\GitHub\corert\src\ILCompiler\src\Program.cs:line 732
```

yowl replied:
I'll do some more investigation here as this came up in the dev here https://github.com/dotnet/corert/issues/5842#issuecomment-430605137 and in your response.

yowl replied:
Its not quite the same, this time the method with no `SavedVertex` is `{[S.P.CoreLib]Internal.Runtime.CompilerServices.Unsafe.As<uint8,object>(uint8&)}` with the generic definition `{[S.P.CoreLib]Internal.Runtime.CompilerServices.Unsafe.As<TFrom,TTo>(!!0&)}`

MichalStrehovsky replied:
Let's make this condition into `if (factory.TypeSystemContext.IsUnboxingThunk(_method))`. The `IsUnboxingThunk` method doesn't exist yet, but there's `IsSpecialUnboxingThunk`.

The implementation will be the same except of the typeof check (instead of `GenericUnboxingThunk`, use `UnboxingThunk`).

MichalStrehovsky replied:
Or maybe wait a bit. I'll build a copy of what you have and look at all the asserts under a debugger. It will be more efficient that way.

MichalStrehovsky commented on src/ILCompiler.Compiler/src/Compiler/MetadataManager.cs:
```diff
@@ -182,13 +182,18 @@ protected virtual void Graph_NewMarkedNode(DependencyNodeCore<NodeFactory> obj)
 
             if (methodNode != null)
             {
+                if (methodNode.Method.OwningType.GetTypeDefinition() is INonEmittableType && IsMethodEligibleForTracking(methodNode.Method))
```
I'm a bit worried about this line.

Is it possible the need for this if block comes from the metadata policy needing tweaks? Do we still need it after addressing the other feedback?

yowl replied:
I tried to remove this one after addressing the other feedback and it hits the assert at https://github.com/dotnet/corert/blob/74897d4ab3eaff1141f8ac07287ccefbfbb5a352/src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/NodeFactory.cs#L537 with stack
```
         Assertion Failed

            at ILCompiler.DependencyAnalysis.NodeFactory.NecessaryTypeSymbol(TypeDesc type) in D:\GitHub\corert\src\ILCompiler.Compiler\src\Compiler\DependencyAnalysis\NodeFactory.cs:line 538
            at ILCompiler.DependencyAnalysis.ExactMethodInstantiationsNode.GetData(NodeFactory factory, Boolean relocsOnly) in D:\GitHub\corert\src\ILCompiler.Compiler\src\Compiler\DependencyAnalysis\ExactMethodInstantiationsNode.cs:line 68
            at ILCompiler.DependencyAnalysis.WebAssemblyObjectWriter.EmitObject(String objectFilePath, IEnumerable`1 nodes, NodeFactory factory, WebAssemblyCodegenCompilation compilation, IObjectDumper dumper) in D:\GitHub\corert\src\ILCompiler.WebAssembly\src\CodeGen\WebAssemblyObjectWriter.cs:line 772
            at ILCompiler.WebAssemblyCodegenCompilation.CompileInternal(String outputFile, ObjectDumper dumper) in D:\GitHub\corert\src\ILCompiler.WebAssembly\src\Compiler\WebAssemblyCodegenCompilation.cs:line 54
            at ILCompiler.Compilation.ILCompiler.ICompilation.Compile(String outputFile, ObjectDumper dumper) in D:\GitHub\corert\src\ILCompiler.Compiler\src\Compiler\Compilation.cs:line 387
            at ILCompiler.Program.Run(String[] args) in D:\GitHub\corert\src\ILCompiler\src\Program.cs:line 576
            at ILCompiler.Program.Main(String[] args) in D:\GitHub\corert\src\ILCompiler\src\Program.cs:line 732
```

yowl commented on src/ILCompiler.Compiler/src/Compiler/CompilerTypeSystemContext.BoxedTypes.cs:
```diff
@@ -107,20 +107,21 @@ public MethodDesc GetUnboxingThunk(MethodDesc targetMethod, ModuleDesc ownerModu
-                return thunkDefinition;
+                thunk = thunkDefinition;
             }
+            //TODO: this might be triggered by a struct that implements an interface with a generic method
```
Deleted

yowl commented on src/ILCompiler/src/Program.cs:
```diff
@@ -7,7 +7,7 @@
 using System.CommandLine;
 using System.Runtime.InteropServices;
-
+using ILCompiler.Compiler;
```
Deleted

yowl commented:
Something in the merge with the ThreadStatic change is causing some parameter overwriting, investigating.

yowl commented:
Fixed up the arg overwriting issue for methods with exception regions and added a test for it.  Could remove the WIP if we're not anticipating anything from @morganbr ,
